src/metabase/api/dashboard.clj

```diff
@@ -215,14 +216,97 @@
+            (cond
+              (nil? (:card_id db-card)) ;; text card
+              []
+
+              (mi/can-write? card)
+              (let [{writable true unwritable false} (group-by (comp boolean mi/can-write?)
+                                                               series)]
+                [(into [card] writable) unwritable])
+              ;; if you can't write the base, we don't have anywhere to put the series
+              :else
+              [[] (into [card] series)]))]
+    (reduce (fn [acc db-card]
+              (let [[retain discard] (split-cards db-card)]
+                (-> acc
+                    (update :copy merge (m/index-by :id retain))
+                    (update :discard concat discard))))
+            {:copy {}
+             :discard []}
+            ordered-cards)))
+
+(defn- duplicate-cards
+  "Takes a dashboard id, and duplicates the cards both on the dashboard's cards and dashcardseries. Returns a map of
+  {old-card-id duplicated-card} so that the new dashboard can adjust accordingly."
+  [dashboard dest-coll-id]
+  (let [same-collection? (= (:collection_id dashboard) dest-coll-id)
+        {:keys [copy discard]} (cards-to-copy (:ordered_cards dashboard))]
+    (reduce (fn [m [id card]]
+              (assoc m id
+                     (if (:dataset card)
+                       card
+                       (api.card/create-card!
+                        (cond-> (assoc card :collection_id dest-coll-id)
+                          same-collection?
+                          (update :name #(str % " -- " (tru "Duplicate"))))))))
+            {:uncopied discard}
+            copy)))
+
+(defn update-cards-for-copy
+  "Update ordered-cards in a dashboard for copying. If shallow copy, returns the cards. If deep copy, replaces ids with
+  id from the newly-copied cards. If there is no new id, it means user lacked curate permissions for the cards
+  collections and it is omitted. Dashboard-id is only needed for useful errors."
+  [dashboard-id ordered-cards deep? id->new-card]
+  (when (and deep? (nil? id->new-card))
+    (throw (ex-info (tru "No copied card information found")
+                    {:user-id api/*current-user-id*
+                     :dashboard-id dashboard-id})))
+  (if-not deep?
+    ordered-cards
+    (keep (fn [dashboard-card]
+            (cond
+              ;; text cards need no manipulation
+              (nil? (:card_id dashboard-card))
+              dashboard-card
+
+              ;; if we didn't duplicate, it doesn't go in the dashboard
+              (not (id->new-card (:card_id dashboard-card)))
+              nil
+
+              :else
+              (let [new-id (fn [id]
+                             (-> id id->new-card :id))]
+                (-> dashboard-card
+                    (update :card_id new-id)
+                    (assoc :card (-> dashboard-card :card_id id->new-card))
+                    (m/update-existing :parameter_mappings
```

Comment:
What about the case where the card's query depends on another card, with `{:dataset_query {:query {:source_table "card__1234"}}}`? Perhaps those are treated like models, and we don't want to duplicate them. I just wanted to make sure it was considered.

Reply:
I don't think that's a problem. If they can edit the collection where the nested query is then its good to go. The UI on the individual card will offer to duplicate it, etc. Good call out though.